File: Step 06 LinkedList/Step 6.3 Medium Problem of LL/LeetCode/Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode head1, ListNode head2) {
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;
        int carry = 0;

        // Traverse both linked lists
        while (head1 != null || head2 != null) {
            int sum = carry;

            // Add the values of the current nodes, if they exist
            if (head1 != null) {
                sum += head1.val;
                head1 = head1.next;
            }
            if (head2 != null) {
                sum += head2.val;
                head2 = head2.next;
            }

            // Update the carry and create a new node with the sum % 10
            carry = sum / 10;
            curr.next = new ListNode(sum % 10);
            curr = curr.next;
        }

        // If there is a remaining carry, create a new node for it
        if (carry > 0) {
            curr.next = new ListNode(carry);
        }

        // Return the head of the result linked list
        return dummy.next;
    }
}

File: Step 06 LinkedList/Step 6.3 Medium Problem of LL/Add Two Numbers/Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addNode(ListNode* &x,int ele){
        ListNode* neww = new ListNode(ele);
        if (x==NULL) x = neww;
        else{
            ListNode *t=x;
            while(t->next!=NULL) t = t->next;
            t->next = neww;
        }
        return x;
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int left=0,sum;
        ListNode *t,*head=NULL,*neww;
        while(l1!=NULL && l2!=NULL){
            sum = l1->val+l2->val+left;
            left = sum/10;
            head = addNode(head,sum%10);
            l1 = l1->next;
            l2 = l2->next;
        }
        while(l1!=NULL){
            sum = l1->val + left;
            left = sum/10;
            head = addNode(head,sum%10);
            l1 = l1->next;
        }
        while(l2!=NULL){
            sum = l2->val + left;
            left = sum/10;
            head = addNode(head,sum%10);
            l2 = l2->next;
        }
        if(left>0) head = addNode(head,left);
        return head;
    }
};